### PR TITLE
[Media] Reactification Template

### PR DESCRIPTION
**Pull request by HenriRabalais**
This Media file will serve as a template for the ongoing reactification of LORIS. 

### In this iteration of refactoring:
 1. I removed all uses of jQuery in the Media file
   - This was done by using `window.onload` as well as Javascript's `fetch()` API
 2. I moved Table Headers into React's jurisdiction by declaring them in the index file.
   - If table headers eventually become configurable, this may be something that is queried from a table, fetched and then set as state.
 3. I moved Filter Form Elements into React's jurisdiction by declaring them in the index file.
  - It is still necessary to fetch the necessary options for select elements from the serverside.
 4. Implemented @driusan's FilterableDataTable component which will handle filtering for all modules. 
 5. Refactored FilterForm component. This involved removing a lot of irrelevant code, such as query strings and all uses of jQuery.
 6. Altered StaticDataTable to take on the responsibility of dealing with not displaying hidden headers and data, as well as organizing the data in format that it useful for the `formatColumn()` function.

Let me know if there is any more work that this PR should include, or if there are any suggestions for handling the Issues.

### CODE REVIEW

File: jsx/FilterableDataTable.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      <div>
+        <FilterForm
+          name={this.props.name + '_filter'}
+          id={this.props.name + '_filter_form'}
+          columns={3}
+          filter={this.state.filter}
+          updateFilter={this.updateFilter}
+          clearFilter={this.clearFilter}
+        >
+          {this.props.formElements}
```
Comment (driusan) [inline jsx/FilterableDataTable.js]: would it make sense for this to be `this.props.children` instead?

File: modules/media/jsx/mediaIndex.js [hunk 2]
```diff
@@ -1,119 +1,174 @@
...
-    this.fetchData();
+    this.fetchData()
+      .then(() => this.setState({isLoaded: true}))
+      .catch((error) => console.error(error));
   }
 
   /**
-   * Retrieve data from the provided URL and save it in state
-   * Additionally add hiddenHeaders to global loris variable
+   * Retrive data from the provided URL and save it in state
```
Comment (driusan) [inline modules/media/jsx/mediaIndex.js]: Retrieve was right before

File: modules/media/jsx/mediaIndex.js [hunk 3]
```diff
@@ -1,119 +1,174 @@
...
+   * Additionaly add hiddenHeaders to global loris vairable
```
Comment (driusan) [inline modules/media/jsx/mediaIndex.js]: Additionally was also right before

File: jsx/Form.js [hunk 4]
```diff
@@ -780,7 +780,7 @@ class TextboxElement extends Component {
   }
 
   handleChange(e) {
-    this.props.onUserInput(this.props.name, e.target.value);
+    this.props.onUserInput(this.props.name, e.target.value, this.props.id, 'textbox');
```
Comment (HenriRabalais) [inline jsx/Form.js]: These two changes in `Form.js` are what allow the `onElementUpdate` function within the `FilterForm` component to accept `id` and `type`. There is likely a way of doing this that doesn't necessitate changing Form.js, but I'm thinking that it may be useful to be able to access the `id` and `type` of components `onUserInput` loris-wide.

File: jsx/FilterForm.js [hunk 5]
```diff
@@ -18,164 +18,49 @@ import Panel from 'Panel';
...
-        formElements[fieldName].onUserInput = this.onElementUpdate.bind(null, formElements[fieldName].type);
-        formElements[fieldName].value = this.queryString[queryFieldName];
-      }.bind(this));
-    }
+    const formChildren = React.Children.map(this.props.children, (child) => {
+      return React.cloneElement(
+        child,
+        {
+          ...child.props,
+          value: (this.props.filter[child.props.name]||{}).value,
```
Comment (davidblader) [inline jsx/FilterForm.js]: nitpick but could you put some spaces in `this.props.filter[child.props.name]||{}`

File: jsx/Form.js [hunk 6]
```diff
@@ -328,7 +328,7 @@ class SelectElement extends Component {
       }
     }
 
-    this.props.onUserInput(this.props.name, value);
+    this.props.onUserInput(this.props.name, e.target.value, e.target.id, 'select');
```
Comment (davidblader) [inline jsx/Form.js]: does using `e.target.value` negate the purpose of lines 322-329?

File: jsx/FilterForm.js [hunk 7]
```diff
@@ -18,164 +18,49 @@ import Panel from 'Panel';
...
+   * @param {string} id
+   * @param {string} type
    */
-  onElementUpdate(type, fieldName, fieldValue) {
-    // Make sure key is of string type and value is of string or object type
-    // before sending them to querystring
-    if (typeof fieldName !== 'string' ||
-        (typeof fieldValue !== 'string' && typeof fieldValue !== 'object')) {
-      return;
+  onElementUpdate(name, value, id, type) {
```
Comment (davidblader) [inline jsx/FilterForm.js]: what's the point of passing `id`?

File: jsx/Form.js (hunk 6 above)
Comment (davidblader) [reply jsx/Form.js]: why does this use the `id` of the event target but `TextboxElement` uses the one passed by `props`?

File: modules/media/jsx/mediaIndex.js (hunk 2 above)
Comment (HenriRabalais) [reply modules/media/jsx/mediaIndex.js]: lol

File: jsx/FilterableDataTable.js (hunk 1 above)
Comment (HenriRabalais) [reply jsx/FilterableDataTable.js]: It could be, I don't see why not.

File: jsx/Form.js (hunk 6 above)
Comment (HenriRabalais) [reply jsx/Form.js]: These are good questions.

File: jsx/FilterForm.js (hunk 5 above)
Comment (HenriRabalais) [reply jsx/FilterForm.js]: For sure.

File: modules/media/jsx/mediaIndex.js (hunk 3 above)
Comment (HenriRabalais) [reply modules/media/jsx/mediaIndex.js]: That's what I get when I don't copy and paste

File: modules/media/jsx/mediaIndex.js [hunk 8]
```diff
@@ -1,119 +1,174 @@
...
-          <MediaUploadForm
-            DataURL={`${loris.BaseURL}/media/ajax/FileUpload.php?action=getData`}
-            action={`${loris.BaseURL}/media/ajax/FileUpload.php?action=upload`}
-            maxUploadSize={this.state.Data.maxUploadSize}
-          />
-        </TabPane>
-      );
-    }
+    const tabList = [{id: 'browse', label: 'Browse'}];
+    const uploadTab = () => {
```
[the remark this answers is not in the document]
Comment (HenriRabalais) [reply modules/media/jsx/mediaIndex.js]: If it beneficial when something is going to be conditionally rendered to put it as a function. That way the variable doesn't have to be declared outside the `if` statement like so:
```
    const uploadTab;                                               
      if (loris.userHasPermission('media_write')) {                             
        tabList.push({id: 'upload', label: 'Upload'});                          
        uploadTab = (                                                                
          <TabPane TabId={tabList[1].id}>                                       
            <MediaUploadForm                                                       
              DataURL={`${loris.BaseURL}/media/ajax/FileUpload.php?action=getData`}
              action={`${loris.BaseURL}/media/ajax/FileUpload.php?action=upload`}
              maxUploadSize={this.state.data.maxUploadSize}                     
            />                                                                  
          </TabPane>                                                            
        );                                                                      
      }                                                                         
    };      
```

File: jsx/Form.js (hunk 6 above)
Comment (davidblader) [reply jsx/Form.js]: makes sense

File: jsx/FilterableDataTable.js (hunk 1 above)
Comment (HenriRabalais) [reply jsx/FilterableDataTable.js]: I sent it as `this.props.children`

**Comment by kongtiaowang** [comment]
When I test this PR all the react table have the issues.
react-dom.development.js:14805 The above error occurred in the <StaticDataTable> component:
    in StaticDataTable (created by DynamicDataTable)
    in DynamicDataTable

 Uncaught TypeError: Cannot read property 'indexOf' of undefined

That is why it failed in Travis.

**Comment by kongtiaowang** [comment]
it is passing Travis.

File: jsx/Filter.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ */
+class Filter extends Component {
+  constructor(props) {
+    super(props);
+    this.onElementUpdate = this.onElementUpdate.bind(this);
+  }
+
+  /**
+   * Sets filter object and querystring to reflect values of input fields
```
Comment (zaliqarosli) [inline jsx/Filter.js]: leftover reference to something that was removed

File: modules/media/jsx/mediaIndex.js [hunk 10]
```diff
@@ -1,121 +1,185 @@
...
+     {label: 'Uploaded By', display: true},
+     {label: 'Date Taken', display: true},
+     {label: 'Comments', display: true},
+     {label: 'Date Uploaded', display: true},
+     {label: 'File Type', display: false},
+     {label: 'Cand ID', display: false},
+     {label: 'Session ID', display: false},
+     {label: 'Hide File', display: false},
+     {label: 'Edit Metadata', display: true},
+    ];
```
Comment (HenriRabalais) [inline modules/media/jsx/mediaIndex.js]: `fields` and `headers` were placed as separate arrays because that's the way things had previously been setup. However, they don't need to be and can likely be placed in to a single array.

File: modules/media/jsx/mediaIndex.js [hunk 11]
```diff
@@ -1,121 +1,185 @@
...
+  /**
+   * Modify behaviour of specified column cells in the Data Table component
+   *
+   * @param {string} column - column name
+   * @param {string} cell - cell content
+   * @param {object} row
+   *
+   * @return {*} a formated table cell for a given column
+   */
+  formatColumn(column, cell, row) {
```
Comment (xlecours) [inline modules/media/jsx/mediaIndex.js]: This is just a comment, an objective one: I prefer when functions have a single return statement.

File: modules/media/php/media.class.inc [hunk 12]
```diff
@@ -108,31 +109,23 @@ class Media extends \NDB_Menu_Filter
...
+        $hiddenOptions = [];
         if ($this->hasHidePermission) {
-            // Unable to create checkbox using addCheckbox(), using addSelect for
-            // the time being
-            // select options
             $hiddenOptions = array(
-                              ""  => 'Both',
-                              "0" => 'Visible only',
-                              "1" => 'Hidden only',
+                              0 => 'Visible only',
```
Comment (xlecours) [inline modules/media/php/media.class.inc]: To display both, one needs to clear the filters?

Comment (HenriRabalais) [reply modules/media/php/media.class.inc]: Clear the filter or return it the blank value that already exists by default in the select dropdown.

File: jsx/DataTable.js [hunk 13]
```diff
@@ -0,0 +1,579 @@
...
+
+    let hasFilters = (filterValuesCount !== 0);
+    if (filterMatchCount === 0 && hasFilters) {
+      return 0;
+    }
+
+    return (filterMatchCount === 0) ? tableData.length : filterMatchCount;
+  }
+
+  toCamelCase(str) {
```
[the remark this answers is not in the document]
Comment (HenriRabalais) [reply jsx/DataTable.js]: As of now, yes. But I am trying to get rid of it in my next commit.

[the remark this answers is not in the document]
Comment (HenriRabalais) [reply jsx/DataTable.js]: Done!

File: modules/media/jsx/mediaIndex.js (hunk 11 above)
Comment (HenriRabalais) [reply modules/media/jsx/mediaIndex.js]: Okay, I will look into making that happen!

File: jsx/Filter.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  height: '100%',
+  title: 'Selection Filter',
+  clearFilter: function() {
+    console.warn('onUpdate() callback is not set!');
+  },
+};
+Filter.propTypes = {
+  filter: PropTypes.object.isRequired,
+  clearFilter: PropTypes.func.isRequired,
+  id: PropTypes.string.isRequired,
```
Comment (driusan) [inline jsx/Filter.js]: why is this required?

File: jsx/FilterableDataTable.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  }
+}
+
+FilterableDataTable.defaultProps = {
+  columns: 3,
+};
+
+FilterableDataTable.propTypes = {
+  name: PropTypes.string.isRequired,
+  children: PropTypes.element.isRequired,
```
Comment (driusan) [inline jsx/FilterableDataTable.js]: children is required but doesn't seem to be used anywhere in this code or passed to a sub-component

File: jsx/FilterableDataTable.js [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+  data: PropTypes.object.isRequired,
+  headers: PropTypes.object.isRequired,
```
Comment (driusan) [inline jsx/FilterableDataTable.js]: same problem as children

File: php/libraries/NDB_Menu_Filter.class.inc [hunk 17]
```diff
@@ -764,8 +764,9 @@ class NDB_Menu_Filter extends NDB_Menu
         }
 
         return array(
-                'Headers' => $headers,
-                'Data'    => $MappedData,
+                'Headers'      => $headers,
+                'Data'         => $MappedData,
+                'fieldOptions' => $this->fieldOptions,
```
Comment (driusan) [inline php/libraries/NDB_Menu_Filter.class.inc]: I'm still not crazy about this mixing of concerns for presentation and data, but this PR is still a step forward so I'm not going to hold it up over this..

Comment (HenriRabalais) [reply php/libraries/NDB_Menu_Filter.class.inc]: I think in the next round of refactoring we can move to request the options in a separate call.

**Comment by kongtiaowang** [comment]
@HenriRabalais 
Fix [cnv_browser.class.inc cpg_browser.class.inc snp_browser.class.inc] in the same way add [public $fieldOptions;]
